test/e2e/dra/test-driver/deploy/example/plugin-rbac.yaml

```diff
@@ -0,0 +1,36 @@
+# Real driver deployments must replace all occurrences of "dra-kubelet-plugin"
+# with something specific to their driver.
+
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: dra-kubelet-plugin-service-account
+  namespace: default
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: dra-kubelet-plugin-role
+rules:
+- apiGroups: ["resource.k8s.io"]
+  resources: ["resourceclaims"]
+  verbs: ["get"]
+- apiGroups: [""]
+  resources: ["nodes"]
+  verbs: ["get"]
+- apiGroups: ["resource.k8s.io"]
+  resources: ["resourceslices"]
+  verbs: ["get", "list", "watch", "create", "update", "patch", "delete" ]
```

Comment:
did this PR punt on creating a validating admission policy that would actually make these write grants safe?

Reply:
Yes. It's tracked in https://github.com/kubernetes-sigs/dra-example-driver/issues/48. Someone wanted to work on it, but there hasn't been much progress. I can look into it.

Reply:
https://github.com/kubernetes/kubernetes/pull/125163#discussion_r1619099182:
> The VAP part scares me because I don't know it and the example seemed not at all simple.

I understand it a bit better now, so it's less scary. I reviewed https://github.com/kubernetes/kubernetes/pull/124711 and left some suggestions on how to improve the test, but overall I think I get it and managed to come up with a similar E2E test for ResourceSlices and DRA drivers. See the last commit in https://github.com/kubernetes/kubernetes/pull/125488 ("DRA e2e: restrict DRA driver permissions with VAP"), if you are interested.

I could pull that commit into this PR, but prefer not to because:
- This PR blocks https://github.com/kubernetes/kubernetes/pull/125488. The sooner we merge it the better.
- The test in this PR would have to use the old API and then needs to be updated for the new one.
- As discussed during the KEP review, we only need to be sure that such a VAP can be written. It is not part of Kubernetes itself.

Speaking of "can be written": it can be written, but as @deads2k points out, "you cannot restrict DELETEs":

https://github.com/kubernetes/kubernetes/blob/f454764fd2071024bf425f6bc11777ffe131151e/test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml#L11-L12

I kind of expected that the node authorizer would apply to the DRA driver plugin, but it doesn't according to my new E2E test. Is it intentionally limited exactly to kubelet? I've not dug into https://github.com/kubernetes/kubernetes/blob/7693a7e71a852009cc114b31c6c37e7a6eb74cb6/plugin/pkg/auth/authorizer/node/node_authorizer.go#L103 but I suppose that is where it bails out. I only get log output further down which shows that it checks kubelet.

DRA driver kubelet plugins want the ability to delete ResourceSlices because the number of such objects associated with the node may have to shrink while the driver runs. I see two options:
- Extend the node authorizer so that it applies to non-kubelet clients.
- Never delete ResourceSlices for a node. Instead, they can be updated to be empty. In practice, that should happen only rarely and if it does, the runtime overhead will probably be low. If the driver exits, kubelet cleans up after them. It has permission to do so (tested in an E2E test).

Neither of these belong into this PR.

Reply:
> I kind of expected that the node authorizer would apply to the DRA driver plugin, but it doesn't according to my new E2E test. Is it intentionally limited exactly to kubelet?

Yes. Finding a way to opt a service account into having its permissions intersected by node authorizer / node admission is on sig-auth's backlog but not in progress yet.

> Speaking of "can be written": it can be written, but as @deads2k points out, "you cannot restrict DELETEs":

What? Admission can definitely restrict deletes...

> I think I get it and managed to come up with a similar E2E test for ResourceSlices and DRA drivers

I expected to see a VAP/VAPBinding object in this PR in the same manifest that added the RBAC permissions that tightened down the broad permissions. I thought that's all that would be required. I essentially don't want to give any examples to people of granting broad RBAC permissions without also locking them down appropriately.

Something like this:

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "dra-kubelet-plugin-node-name-match"
spec:
  failurePolicy: Fail
  matchConstraints:
    resourceRules:
      - apiGroups:   ["resource.k8s.io"]
        apiVersions: ["*"]
        operations:  ["CREATE", "UPDATE", "DELETE"]
        resources:   ["resourceslices"]
  matchConditions:
    - name: isRestrictedUser
      expression: request.userInfo.username == 'system:serviceaccount:default:dra-kubelet-plugin-service-account'
  variables:
    - name: userNodeName
      expression: request.userInfo.extra[?'authentication.kubernetes.io/node-name'][0].orValue('')
    - name: objectNodeName
      expression: "(request.operation == 'DELETE' ? oldObject : object).?nodeName.orValue('')"
  validations:
    - expression: variables.userNodeName == ''
      message: "this user must have a \"authentication.kubernetes.io/node-name\" claim by running in a pod on a node"
    - expression: variables.userNodeName != variables.objectNodeName
      message: "this user may only modify resourceslices that belong to the node the pod is running on"
---
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicyBinding
metadata:
  name: "dra-kubelet-plugin-node-name-match"
spec:
  policyName: "dra-kubelet-plugin-node-name-match"
  validationActions: [Deny]
```

> > Never delete ResourceSlices for a node.
> 
> This seems like a very minor limitation that doesn't hurt at all in practice. I went ahead and implemented that in another commit in #125488 ("DRA kubelet: avoid DELETE permission for ResourceSlices").

I'm not sure I see why that's needed... what you had here for the kubelet seemed ok... allow deletes, ensure in admission that the kubelet is only deleting ones for itself.

Reply:
> I expected to see a VAP/VAPBinding object in this PR in the same manifest that added the RBAC permissions that tightened down the broad permissions.

That's what I am doing in the follow-up. Okay, so let's pull that in here.

> What? Admission can definitely restrict deletes...

Hmm, not sure why @deads2k writes that it can't. I'll check.

Reply:
https://github.com/kubernetes/kubernetes/pull/124711/files#r1683339122

Reply:
When using `oldObject` it works. See the [dra e2e: demonstrate how to use RBAC + VAP for a kubelet plugin](https://github.com/kubernetes/kubernetes/pull/125163/commits/80bb25afef465d20841f231d6644097cccf81ec7) commit.